### ISSUE TITLE
shmem_internal_team_release_psyncs impl is broken

### ISSUE THREAD

**Issue by akhillanger** [Sandia-OpenSHMEM/SOS#963]
For SYNC op, there is only one array and that can used back-to-back w/o waiting for prior SYNC ops to complete.
For REDUCE, you have more than one psync array per team and it is chosen selectively using choose_psync fn. Hence you need to selectively release them also.

Also, I see that there is no release_psync fn corresponding to choose_psync at line - https://github.com/Sandia-OpenSHMEM/SOS/blob/master/src/shmem_team.c#L277

Tagging @davidozog and @jdinan

**Comment by davidozog** [comment]
@akhillanger Wow, good catch.  IIUC, it looks like releasing psyncs for REDUCE/BCAST/COLLECT/ALLTOALL doesn't actually do _anything_... This must have been an oversight on my part. 🤦  But SYNC is ok, right?

Off the cuff, it seems like it might help to carry around an index to help facilitate selectively choosing/releasing.  Something like:
```
long* choose_psync(shm_team_t *team, shm_team_op_t op, size_t *idx)
void release_psyncs(shm_team_t *team, shm_team_op_t op, const size_t idx)
```
where `choose` sets `*idx` and `release` resets `team->psync_avail[idx]` to 1.  `idx` isn't needed with SYNC, so could be NULL.  Seems slightly kludgy, so perhaps there's a better approach you can think of?

And yeah, after we fix this there should be a release corresponding to shmem_team.c#L277

Please feel free to post an alternative solution.  Sounds like you might be more familiar with the code than I am at the moment - I'm having to remind myself of many things...

**Comment by akhillanger** [comment]
> @akhillanger Wow, good catch. IIUC, it looks like releasing psyncs for REDUCE/BCAST/COLLECT/ALLTOALL doesn't actually do _anything_... This must have been an oversight on my part. 🤦 But SYNC is ok, right?
> 
Yes, that is correct David. Just ensure that your sync implementation is ok using the same psync array for back-to-back sync collective.
> Off the cuff, it seems like it might help to carry around an index to help facilitate selectively choosing/releasing. Something like:
> 
> ```
> long* choose_psync(shm_team_t *team, shm_team_op_t op, size_t *idx)
> void release_psyncs(shm_team_t *team, shm_team_op_t op, const size_t idx)
> ```
> 
> where `choose` sets `*idx` and `release` resets `team->psync_avail[idx]` to 1. `idx` isn't needed with SYNC, so could be NULL. Seems slightly kludgy, so perhaps there's a better approach you can think of?
> 
I haven't thought of anything better yet.
> And yeah, after we fix this there should be a release corresponding to shmem_team.c#L277
> 
> Please feel free to post an alternative solution. Sounds like you might be more familiar with the code than I am at the moment - I'm having to remind myself of many things...

**Comment by davidozog** [comment]
I implemented the patch (along with the bug fixes in #960) but now see some deadlocks in certain collective algorithms with back-to-back calls in the `shmemx_team_b2b_collectives` unit test.

Both `collect` and `fcollect` deadlock early (after ~2-3 iterations) with only 2 PEs.  1 PE is in the final `shmem_internal_barrier` (using the same `psync` as chosen by `choose_psync(COLLECT,...)`), and the other PE is in a `wait_until` waiting for payload data.

The **tree** `bcast` also deadlocks in a similar fashion after 3-4 iterations with at least 7-8 PEs.  The problem goes away if I force a linear bcast. 

Back-to-back barriers and syncs seem fine, and so do reduce, alltoall, and alltoalls.

Any ideas about what's causing this?  Thanks in advance for any help you can offer @akhillanger, @jdinan, @wrrobin.  (no problem if not - this seems like a doozy).

Here's the diff/branch in case you have a moment to check for mistakes:
https://github.com/Sandia-OpenSHMEM/SOS/compare/master...davidozog:wip/teams_bug_fixes?expand=1

**Comment by akhillanger** [comment]
On a slightly different note, I wanted to ask in what scenario do you need more than one psync array? Since we only have blocking collectives there can be only one outstanding collective ay any time. And the psync will be released once the collective is done.

Multiple psync arrays per team will be useful only for non-blocking collectives.

**Comment by davidozog** [comment]
We had back-to-back collectives in mind when deciding to support multiple psync arrays per team.  Passing separate psync arrays to back-to-back collectives is a common application-level optimization (see the v5.6.1 OSU broadcast/other microbenchmarks, for example).  SOS doesn't yet use them unless for some reason the user is trying to call a collective on the same team from different threads (which is not allowed and would possibly explode).  Having multiple psync arrays per team _available_ is a step in the direction of supporting multi-threaded and/or non-blocking collectives.  I suppose we could/should set `N_PSYNCS_PER_TEAM` to 1... but having 2 is probably not too expensive.

I thought this was an optimization choice though, not a correctness issue, but @jdinan please correct me if I'm wrong.  I hope that's not the bug above... 😰

**Comment by davidozog** [comment]
Well... that was a great question you asked, @akhillanger, because it does seem to be directly related to the bug.  We might want to disregard a bit of what I said above about back-to-back blocking collectives.

I can prove in a debugger, for example with 2 PEs, that both PEs can reach the the final synchronization in `shmem_internal_collect_linear` (the `shmem_internal_barrier` towards the end of that routine), and then 1 PE can race way ahead of the other, getting all the way to the next collective on the same chosen `psync` _before_ the other PE has successfully reset its `psync` values to `SHMEM_SYNC_VALUE`, causing a deadlock.

If I make use of the other psync arrays that are available to the team, that race goes away, and I can convince myself that the race cannot occur when swapping between only 2 psync arrays per team (and I think that's the case for all SOS's collectives algorithms).  I seem to recall @jdinan predicting this would happen a long time ago,  but with the bugs that were in `release_psync`, it never came up until now.

Does this all sound correct?  If so, I'll post a PR soon with all the teams fixes.

**Comment by akhillanger** [comment]
This sounds correct David. However, I don't think this will work across all combinations of collective algorithms. Consider a scenario, recursive_doubling(algo 1)->tree based bcast(algo 2)->recursive doubling(algo 3). It's possible that a rank (say root rank 0) has moved to algorithm 3, while one of its peers is still in algorithm 1 that uses the same psync array as algorithm 1. Please verify if my thinking is correct. The solution here would be to have two psync arrays just for each type of algo maybe.


If we consider the active-set based collectives OpenSHMEM API in which the user provides the pSync array as input to the API, the issue that you mentioned above exists. Hence, I think the logic of ensuring that the pSync value has been reset has to be part of the algorithm itself. Before communicating in an algo, peer PEs have to first exchange hand-shake that they are ready to receive the data, basically telling that we have exited the previous collective call. With this approach, we can do with just one psync but it has this additional hand-shake overhead.

**Comment by akhillanger** [comment]
Thinking more - it seems that single pSync solution will not work even with hand-shake. And I think, therefore for active-set based collectives API, OpenSHMEM also would require a barrier between two calls to collective ops for ensuring that pSync has been reset on all PEs before moving on to next collective call.

And in team-based API, the solution would be to have two psyncs. But based on the reason I gave in para 1 above, we would probably need two psyncs for each type of algorithm.

**Comment by davidozog** [comment]
Agreed - I'm convinced a single psync array cannot work for b2b collectives (in general;) unless there is a barrier between the collectives at the application level.

I think we need at least 2 psyncs per team and yes, there may be different synchronization requirements across different collective operations and algorithms.  I'm in support of a case statement across algorithms someday (please help if you have the time!).  We could also make N_PSYNCS_PER_TEAM adjustable or a part of shmemx_team_config_t...  Lots of options - any preferences?

**Comment by jdinan** [comment]
@davidozog Doesn't SOS insert a barrier between collectives in the single pSync case you mentioned?

**Comment by davidozog** [comment]
@jdinan Do you mean at the lowest level in `collectives.c` or somewhere else higher up?  I suppose it's possible I missed a barrier high up, but I can't find it... just choosing/releasing psyncs, which are local ops.

In the case of a single pSync per team, if we look at `shmem_internal_collect_linear` for example (where the bug is easy to reproduce) there is an explicit `shmem_internal_barrier` at the very end of the function.  If I have a debugger with 2 PEs, I can synchronize immediately after that barrier (before resetting the pSync), then allow the 1st PE to carry all the way through to the next collect_linear call, and it eventually arrives at that same barrier (since it's the root), all before the 2nd PE has even had a chance to reset the pSync array.

Does that help explain?  I have a few more bold claims, but I want to debug a bit more before going into details.  I'll double check the above claim too.

**Comment by davidozog** [comment]
I double-checked the above, and it's definitely reproducible (at least with @akhillanger's bug fixes included, without them all bets are off ;).

I'm able to verified another bold claim empirically, but can't make much sense of it yet....

With 1 pSync array per team, even putting an explicit internal barrier at the top of `shmem_internal_collect_linear` still runs into deadlock with a sufficient # of PEs (I can catch it with 12 PEs on a laptop occasionally).  It's certainly far less frequent, but can still happen...

However, with a barrier between b2b collectives at the application level, I so far cannot produce a deadlock.  Maybe this is incidental and not actually a property of b2b collectives in general.  I'm trying to figure that out... it kinda makes sense that assuring the pSync is reset across all PEs before entering the collective help to avoid these issues with a single pSync.

Anyway, if I swap between multiple available pSync arrays on a team between collectives, then it seems we don't need a barrier between b2b collectives in practice.  But it's hard to prove in general because it probably depends on the collective and it's algorithm, as @akhillanger claims.

Unless I hear otherwise, I'll post a PR with that makes use of the 2 pSync arrays and we can discuss any concerns on the ticket.

**Comment by jdinan** [comment]
IIRC, pSync selection should be collective. If an inactive pSync is not available, then a barrier is needed to ensure pSyncs are no longer in use.

**Comment by davidozog** [comment]
@jdinan - As of now, SOS w/teams does a barrier during pSync selection only if no inactive pSyncs are available.  I believe that barrier needs to be there.  But are you saying there needs to be an additional barrier even if a pSync is available?  I might need to be convinced of that... it's a lot of barriers 😉.  With b2b active-set based collectives, you need a barrier at the application level to reuse a single pSync, right?

**Comment by jdinan** [comment]
The barrier during pSync selection should prevent pSync races in back-to-back collectives on the same team, right?

**Comment by davidozog** [comment]
Ah - so if by "The barrier" you mean the one that occurs when no inactive pSync is available, I think it would only prevent the race if all `N_PSYNCS_PER_TEAM` pSync arrays were actually being utilized.  In the current implementation on master, only the first pSync array is ever used, so "The barrier" is never actually called with this behavior/bug.

I have a PR about ready that iterates through all (2) `N_PSYNCS_PER_TEAM` and fixes this issue (I think).

**Comment by jdinan** [comment]
The `release_psyncs` function should only mark psyncs as available if they can be used immediately. The only case where this can be true is when the `op` is `SYNC`. That is, if `op` was a reduction or broadcast, you can't use that psync again until there is a barrier.

I do see a potential race caused by the barriers not quiescing collective communication. This looks like an issue in `choose_psync` (calls `shmem_internal_sync` to clear psyncs) and also in `release_psync` since `shmem_team_sync` does not guarantee that collective communication is quiesced prior to the process synchronization.

**Comment by davidozog** [comment]
> The `release_psyncs` function should only mark psyncs as available if they can be used immediately. The only case where this can be true is when the `op` is `SYNC`. That is, if `op` was a reduction or broadcast, you can't use that psync again until there is a barrier.

Yeah, that's a good point - thanks for catching it.

> I do see a potential race caused by the barriers not quiescing collective communication. This looks like an issue in `choose_psync` (calls `shmem_internal_sync` to clear psyncs) and also in `release_psync` since `shmem_team_sync` does not guarantee that collective communication is quiesced prior to the process synchronization.

Oh yeah, so you're saying we should replace that with `shmem_internal_barrier`?  That seems promising.

**Comment by davidozog** [comment]
I played around with some b2b collective tests, and cannot reproduce the deadlock with @jdinan's suggestion.  I think this issue may be something of a false alarm brought about by the assumption that we need to selectively release the pSync array for non-SYNC collectives.

Additionally, I cannot produce any problems when using `shmem_internal_sync` in `choose_psync` and doing no quiescing in `release_psync`, but it seems like a valid concern.  I'll think about it, test some more, and hopefully post a PR next week.

**Comment by jdinan** [comment]
SOS is still using the default context on the world team for collectives on all teams, right? I would continue to use sync and add a quiet of the context first with a strongly worded comment that it *must* quiesce any communication involving the `psync`.

**Comment by davidozog** [comment]
> SOS is still using the default context on the world team for collectives on all teams, right? 

@jdinan - yes indeed, though that may not be the case for long (I hope).

> I would continue to use sync and add a quiet of the context first with a strongly worded comment that it _must_ quiesce any communication involving the `psync`.

:+1:

**Comment by davidozog** [comment]
Closed by #968
